src/internal/consistenthashing/etcd_test.go

```diff
@@ -0,0 +1,294 @@
+			test.runWorker(test.ctx, t, nodeId)
+			return nil
+		})
+	}
+	test.lockAllLocks()
+	test.workers++
+	memberToAdd := strconv.Itoa((test.workers - 1) * 100)
+	test.workerIds[test.workers-1] = memberToAdd
+	test.beginLocking <- true
+	test.eg.Go(func() error {
+		test.runWorker(test.ctx, t, memberToAdd)
+		return nil
+	})
+	rebalancedLocks := test.locks / test.workers
+	for i := 0; i < rebalancedLocks; i++ {
+		<-test.doneLocking
+	}
+	locksPerMember := test.locksPerWorker()
+	require.Equal(t, locksPerMember[memberToAdd], rebalancedLocks)
+	test.unlockAllLocks()
+}
+
+// This hash function should get us pretty close to an even distribution,
+// given how lock tests define our key and node names.
+func hashFnForTests(t *testing.T) func([]byte) uint32 {
+	return func(key []byte) uint32 {
+		key = []byte(path.Base(string(key)))
+		keyAsInt, err := strconv.Atoi(string(key))
+		require.NoError(t, err, "conversion should succeed")
+		return uint32(keyAsInt)
+	}
+}
+
+func setupLockTest(t *testing.T, numNodes, numLocks int) lockTestConfig {
+	hashFn = hashFnForTests(t)
+	// number of goroutines is test.Nodes * test.locks, so we should test with fairly small numbers.
+	test := lockTestConfig{
+		locks:      numLocks,
+		keys:       &sync.Map{},
+		workers:    numNodes,
+		workerIds:  map[int]string{},
+		ringConfig: setupTest(t),
+	}
+	test.eg, test.ctx = errgroup.WithContext(test.ringConfig.ctx)
+	test.ctx, test.cancel = context.WithCancel(test.ctx)
+	test.workersReady = make(chan bool, numNodes)
+	test.beginLocking = make(chan bool, numNodes)
+	test.doneLocking = make(chan bool, numLocks)
+	test.beginUnlocking = make(chan bool, numLocks)
+	test.doneUnlocking = make(chan bool, numLocks)
+	// starting from 1 is intentional here, allows the lock distribution to be even.
+	for i := 1; i <= test.locks; i++ {
+		key := strconv.Itoa(i%test.workers*100 + i)
+		test.keys.Store(key, "")
+	}
+	for i := 0; i < test.workers; i++ {
+		test.workerIds[i] = strconv.Itoa(i * 100)
+	}
+	return test
+}
+
+func (test *lockTestConfig) lockAllLocks() {
+	for i := 0; i < test.workers; i++ {
+		<-test.workersReady
+	}
+	for i := 0; i < test.workers; i++ {
+		test.beginLocking <- true
+	}
+	for i := 0; i < test.locks; i++ {
+		<-test.doneLocking
+	}
+}
+
+func (test *lockTestConfig) unlockAllLocks() {
+	for i := 0; i < test.locks; i++ {
+		test.beginUnlocking <- true
+	}
+	for i := 0; i < test.locks; i++ {
+		<-test.doneUnlocking
+	}
+}
+
+func (test *lockTestConfig) locksPerWorker() map[string]int {
+	locksPerWorker := map[string]int{}
+	for _, worker := range test.workerIds {
+		locksPerWorker[worker] = 0
+	}
+	test.keys.Range(func(k any, v any) bool {
+		node := v.(string)
+		locksPerWorker[node]++
+
+		return true // return true acts as continue in a sync.Map.Range()
+	})
+	return locksPerWorker
+}
+
+func (test *lockTestConfig) runWorker(ctx context.Context, t *testing.T, id string) {
+	collection.DefaultTTL = 1
+	idGen = func() string { return id }
```

Comment:
This seems like an issue since this function is running in a goroutine. A different value could be set for idGen before it is used in `WithRing`.

Reply:
Any ideas on how I can do a dependency injection of an Id for testing?